Add technical writer edits (#1)
OVI3D0/project-website: _posts/2025-04-24-Redline-Testing-Comes-to-OpenSearch-Benchmark.md

---
layout: post
title: Redline testing now available in OpenSearch Benchmark
authors: Michael Oviedo
date: 2025-04-24 12:30:00 -0600
categories:
  - technical-posts
meta_keywords: OpenSearch Benchmark, OpenSearch Benchmark, workloads, redline testing
meta_description: Learn how the new redline testing feature in OpenSearch Benchmark helps you automatically determine the maximum throughput your cluster can handle—no guesswork required.
---

# Redline testing now available in OpenSearch Benchmark

> **Automatically determine your OpenSearch cluster’s throughput limits—no guesswork required**

Reviewer: ```suggestion
**Automatically determine your OpenSearch cluster's throughput limits—no guesswork required**
```

> “How much traffic can my cluster really handle?” is a question many OpenSearch users face before moving to production. Until now, finding your cluster’s **redline**—the point beyond which it can no longer maintain acceptable service levels—required trial and error, guesswork, or time-consuming manual tuning.

Reviewer: ```suggestion
"How much traffic can my cluster really handle?" is a question many OpenSearch users face before moving to production. Until now, finding your cluster's **redline**—the point beyond which it can no longer maintain acceptable service levels—required trial and error, guesswork, or time-consuming manual tuning.
```

> With redline testing in OpenSearch Benchmark , you can now dynamically scale client load based on real-time cluster performance to automatically identify your cluster’s throughput ceiling.

Reviewer: ```suggestion
With redline testing in OpenSearch Benchmark, you can now dynamically scale client load based on real-time cluster performance to automatically identify your cluster's throughput ceiling.
```

## The challenge of guessing thresholds

OpenSearch Benchmark recently introduced support for ramping up clients and throughput during a benchmark run. However, users still needed to perform the following actions when benchmarking:

- Estimate what amount of load would break the cluster.  
- Manually watch for failures in logs or dashboards.  
- Rerun benchmarks with different parameters.  

These time-consuming tasks made it difficult to pinpoint the maximum sustainable throughput.

## What we built: Redline testing

Redline testing automates the following:

- A **Feedback Actor** to monitor request failures in real time  
- A **self-adjusting load mechanism** to pause or unpause clients based on observed behavior  

> This enables OpenSearch Benchmark to perform following actions:

Reviewer: ```suggestion
This enables OpenSearch Benchmark to perform the following actions:
```

- Ramp up active clients  
- Detect when failures begin  
- Scale back automatically, wait for recovery, and resume testing  

> The result: OpenSearch Benchmark can now determine your cluster’s redline in a single test run.

Reviewer: ```suggestion
The result: OpenSearch Benchmark can now determine your cluster's redline in a single test run.
```

## How it works

The following diagram provides a high-level overview of the actor-based execution flow in OpenSearch Benchmark. It illustrates how benchmark tasks are allocated and executed across multiple worker actors, each of which manages a group of clients responsible for sending requests to a target OpenSearch cluster, using the following components:

- **BenchmarkActor**: Initiates the benchmarking process.  
- **WorkerCoordinatorActor**: Manages worker lifecycle and task distribution using an allocation matrix from the **Allocator**.  

> - **Workers (Worker1 through WorkerN)**: Execute tasks by managing clients via the **AsyncIoAdapter**.  

Reviewer: ```suggestion
- **Workers (Worker1 through WorkerN)**: Executes tasks by managing clients via the **AsyncIoAdapter**.  
```

- **Clients (Client1 through ClientN)**: Uses the `AsyncExecutor` class to perform operations against the target host in parallel.  

![Flowchart for OpenSearch Benchmark's actor system](/assets/media/blog-images/2025-04-24-Redline-Testing-Comes-to-OpenSearch-Benchmark/OpenSearch Benchmark-system-architecture.jpg){: .img-fluid}

OpenSearch Benchmark uses the **Actor Model**, which structures concurrent, distributed systems around isolated, message-passing components.

Each actor performs the following actions:

- Maintains private, isolated state  
- Processes messages sequentially  
- Can create other actors  
- Communicates asynchronously  

By default, OpenSearch Benchmark spawns `n` worker actors (one per CPU core), distributing clients evenly across them. For example, with 10 workers and 20 clients, each manages 2 clients.
 

### Timed mode and redline logic

OpenSearch Benchmark supports the following two modes for redline testing:

- **Iteration mode**: Runs tasks for a fixed number of iterations  
- **Timed mode**: Runs tasks for a fixed duration  

Redline testing uses timed mode to perform the following actions:

> 1. Ramp up client load until errors occur  

Reviewer: ```suggestion
1. Ramp up client load until errors occur.
```

> 2. Scale down in response to failures  

Reviewer: ```suggestion
2. Scale down in response to failures.
```

> 3. Wait for recovery and ramp up again  

Reviewer: ```suggestion
3. Wait for recovery and ramp up again.
```

### Feedback Actor and shared state

Failures are sent to the **Feedback Actor** via a shared queue. A shared dictionary manages the client state:

```json
{
  "worker-1": { "client-0": true, "client-1": false }
}
```

The following settings reflect these client states: 

- `true` = active  
- `false` = paused  

> This state is shared using Python’s `multiprocessing` module. Each client checks the dictionary before sending a request—paused clients skip execution.

Reviewer: ```suggestion
This state is shared using Python's `multiprocessing` module. Each client checks the dictionary before sending a request—paused clients skip execution.
```

For more technical details, see the [RFC on redline testing](https://github.com/opensearch-project/opensearch-benchmark/issues/785#issue-2898221524).

## Getting started with redline testing

> Getting started with redline testing begins with a timed test procedure that defines the duration, target throughput, and number of clients. Once configured, you can run the benchmark with a single command—and optionally customize the maximum number of clients to match your cluster’s capacity.

Reviewer: ```suggestion
Getting started with redline testing begins with a timed test procedure that defines the duration, target throughput, and number of clients. Once configured, you can run the benchmark with a single command—and optionally customize the maximum number of clients to match your cluster's capacity.
```

  ]
}
```

Then run the benchmark test with the following command:

```bash
opensearch-benchmark execute-test \
  --pipeline=benchmark-only \
  --target-hosts=<your-opensearch-cluster> \
  --workload=<workload> \
  --test-procedure=<your-timed-mode-test-procedure> \
  --redline-test
```

To override the default max clients (1000), add a number to the `--redline-test` flag, similar to the following example:

```bash
--redline-test=1500
```

> This is useful if your cluster can sustain more than 1000 clients for your workload.

Reviewer 2: This will change with https://github.com/opensearch-project/opensearch-benchmark/pull/835

Maybe we can reword it to some version of:
"Users can customize the specifics of the redline test like maximum client count, how quickly the number of clients scales, what percentage of clients are paused, and how long the system sleeps before attempting to scale up again."

> This is useful if your cluster can sustain more than 1000 clients for your workload.

Author: ```suggestion
Users can customize redline test parameters such as the maximum number of clients, the client ramp-up rate, the percentage of clients to pause during back-off, and the wait time before resuming scale-up.
```

> After the test, OpenSearch Benchmark logs the following data:

Author: ```suggestion
During testing, OpenSearch Benchmark logs the following data:
```

- Maximum number of stable clients  

> - All pause and unpause events  

Reviewer 2: these two are logged during the test, not after

> - Reasons for scaling back  

Reviewer 2: Nit: these two are logged during the test, not after

Author: Maybe make it more general: "OpenSearch Benchmark will log the following information?"

Reviewer 2: Yeah that works IMO. To be specific, here's what's logged
During the test:
- Current number of clients
- Pause/unpause events
- Reasons for scaling back

After the test:
- Max number of clients reached
- Performance metrics

Author: Did the best of both worlds here.

  ]
}
```

Then run the benchmark test with the following command:

```bash
opensearch-benchmark execute-test \
  --pipeline=benchmark-only \
  --target-hosts=<your-opensearch-cluster> \
  --workload=<workload> \
  --test-procedure=<your-timed-mode-test-procedure> \
  --redline-test
```

> To override the default max clients (1000), add a number to the `--redline-test` flag, similar to the following example:

Reviewer: ```suggestion
To override the default max clients (1000), add a number to the `--redline-test` flag, as shown in the following example:
```

```bash
--redline-test=1500
```

Users can customize redline test parameters such as the maximum number of clients, the client ramp-up rate, the percentage of clients to pause during back-off, and the wait time before resuming scale-up.

OpenSearch Benchmark captures the following log information:
- During the test:

>   - The current number of clients.

Reviewer: ```suggestion
  - The current number of clients
```

>   - The pause/unpause events.

Reviewer: ```suggestion
  - The pause/unpause events
```

  - The reasons for scaling back
- After the test:
  - The maximim number of clients reached
  - Any performance metrics

> The following chart shows how redline testing in OpenSearch Benchmark incrementally increases client load during a timed test. Each step represents a controlled ramp-up, allowing the system to observe when performance begins to degrade. In this example image, throughput steadily rises until it plateaus—indicating the cluster’s redline has been reached. This automated feedback loop removes guesswork and enables precise load testing in a single run.

Reviewer: ```suggestion
The following chart shows how redline testing in OpenSearch Benchmark incrementally increases client load during a timed test. Each step represents a controlled ramp-up, allowing the system to observe when performance begins to degrade. In this example image, throughput steadily rises until it plateaus—indicating that the cluster's redline has been reached. This automated feedback loop removes guesswork and enables precise load testing in a single run.
```

![Latency over time](/assets/media/blog-images/2025-04-24-Redline-Testing-Comes-to-OpenSearch-Benchmark/dashboards-latency-over-time.png){: .img-fluid}

> ## What’s next?

Reviewer: ```suggestion
## What's next?
```

> Upcoming improvements to redline testing might include:

Reviewer: ```suggestion
Upcoming improvements to redline testing may include:
```

- Smarter ramp-up strategies, such as binary or exponential search.  
- Scaling based on latency or service time, not just request failures.  

> Redline testing is available now in OpenSearch Benchmark. Try it out and share your feedback by making a [GitHub issue](https://github.com/opensearch-project/OpenSearch-Benchmark/issues)!

Reviewer: ```suggestion
Redline testing is available now in OpenSearch Benchmark. Try it out and share your feedback by creating a [GitHub issue](https://github.com/opensearch-project/OpenSearch-Benchmark/issues)!
```
